File: Simple-OTT/app/src/main/java/com/theoplayer/demo/simpleott/view/StreamSourceAdapter.java
package com.theoplayer.demo.simpleott.view;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;

import androidx.annotation.NonNull;

import com.theoplayer.demo.simpleott.PlayerActivity;
import com.theoplayer.demo.simpleott.databinding.LayoutStreamSourceBinding;
import com.theoplayer.demo.simpleott.model.StreamSource;

import java.util.List;

import static android.view.LayoutInflater.from;

public class StreamSourceAdapter extends ArrayAdapter<StreamSource> {

    public StreamSourceAdapter(Context context, List<StreamSource> streamSources) {
        super(context, 0);
        addAll(streamSources);
    }

    @NonNull
    @Override
    public View getView(int position, View rowView, @NonNull ViewGroup parent) {
        LayoutStreamSourceBinding binding;
        if (rowView == null) {
            binding = LayoutStreamSourceBinding.inflate(from(getContext()), parent, false);
            rowView = binding.getRoot();
            rowView.setTag(binding);
        } else {
            binding = (LayoutStreamSourceBinding) rowView.getTag();
        }

        StreamSource streamSource = getItem(position);

        if (streamSource != null) {
            rowView.setOnClickListener(v -> PlayerActivity.play(getContext(), streamSource));
            binding.setViewModel(streamSource);
        }

        return rowView;
    }
}

File: simple-ott/src/main/java/com/theoplayer/demo/simpleott/view/StreamSourceAdapter.kt
package com.theoplayer.demo.simpleott.view

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.theoplayer.demo.simpleott.PlayerActivity
import com.theoplayer.demo.simpleott.databinding.LayoutStreamSourceBinding
import com.theoplayer.demo.simpleott.model.StreamSource

class StreamSourceAdapter(
    private val streamSources: List<StreamSource>
) : RecyclerView.Adapter<StreamSourceAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val viewBinding = LayoutStreamSourceBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return ViewHolder(viewBinding)
    }

    override fun getItemCount(): Int = streamSources.size

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        viewHolder.bind(streamSources[position])
    }

    inner class ViewHolder(private val viewBinding: LayoutStreamSourceBinding) :
        RecyclerView.ViewHolder(viewBinding.root) {

        fun bind(streamSource: StreamSource) {
            viewBinding.sourceImageView.setImageResource(streamSource.imageResId)
            viewBinding.titleTextView.text = streamSource.title
            viewBinding.descriptionTextView.text = streamSource.description
            viewBinding.container.setOnClickListener {
                PlayerActivity.play(
                    viewBinding.root.context,
                    streamSource.source,
                    streamSource.title,
                    streamSource.description
                )
            }
        }
    }
}
